src/VecSim/spaces/computer/preprocessor.h

```diff
@@ -0,0 +1,75 @@
+/*
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+#include <cstddef>
+#include <memory>
+
+#include "VecSim/memory/vecsim_base.h"
+#include "VecSim/spaces/spaces.h"
+#include "VecSim/memory/memory_utils.h"
+
+class PreprocessorAbstract : public VecsimBaseObject {
+public:
+    PreprocessorAbstract(std::shared_ptr<VecSimAllocator> allocator)
+        : VecsimBaseObject(allocator) {}
+    struct PreprocessParams;
+    virtual void preprocess(const void *original_blob, MemoryUtils::unique_blob &storage_blob,
+                            MemoryUtils::unique_blob &query_blob,
+                            PreprocessParams &params) const = 0;
+    virtual void preprocessForStorage(MemoryUtils::unique_blob &blob) const = 0;
+    virtual void preprocessQuery(MemoryUtils::unique_blob &blob) const = 0;
+
+    virtual bool hasQueryPreprocessor() const = 0;
+    virtual bool hasStoragePreprocessor() const = 0;
+
+    struct PreprocessParams {
```

Comment:
Is this the state of the blobs? I'm not sure it should belong to the preprocessor. I'm also not sure why we need to keep it dynamic - aren't we supposed to create the right pipeline and then just use it without validations?

Reply:
The computer flags should determine type of unique ptr (frees the blob at the end or not) and the preprocessor params hint the preprocessor if the unique pte is already populated or needs to be initialized.